File: lib/src/main/java/com/sd/lib/cache/CacheConfig.java
package com.sd.lib.cache;

import android.content.Context;

import com.sd.lib.cache.store.InternalDiskCacheStore;

public class CacheConfig {
    private static CacheConfig sConfig;

    public final Context mContext;
    public final Cache.ObjectConverter mObjectConverter;
    public final Cache.EncryptConverter mEncryptConverter;
    public final Cache.ExceptionHandler mExceptionHandler;

    public final Cache.CacheStore mDiskCacheStore;

    private CacheConfig(Builder builder) {
        mContext = builder.mContext;
        mObjectConverter = builder.mObjectConverter;
        mEncryptConverter = builder.mEncryptConverter;
        mExceptionHandler = builder.mExceptionHandler != null ? builder.mExceptionHandler : new Cache.ExceptionHandler() {
            @Override
            public void onException(Exception e) {
            }
        };

        mDiskCacheStore = builder.mDiskCacheStore != null ? builder.mDiskCacheStore : new InternalDiskCacheStore(mContext);
    }

    /**
     * 初始化
     *
     * @param config
     */
    public static synchronized void init(CacheConfig config) {
        if (config == null) {
            throw new IllegalArgumentException("config is null");
        }

        if (sConfig == null) {
            sConfig = config;
        }
    }

    /**
     * 返回配置
     *
     * @return
     */
    public static CacheConfig get() {
        if (sConfig == null) {
            throw new RuntimeException(CacheConfig.class.getSimpleName() + "has not been init");
        }
        return sConfig;
    }

    public static final class Builder {
        private Context mContext;
        private Cache.ObjectConverter mObjectConverter;
        private Cache.EncryptConverter mEncryptConverter;
        private Cache.ExceptionHandler mExceptionHandler;

        private Cache.CacheStore mDiskCacheStore;

        /**
         * 设置对象转换器
         *
         * @param converter
         * @return
         */
        public Builder setObjectConverter(Cache.ObjectConverter converter) {
            mObjectConverter = converter;
            return this;
        }

        /**
         * 设置加解密转换器
         *
         * @param converter
         * @return
         */
        public Builder setEncryptConverter(Cache.EncryptConverter converter) {
            mEncryptConverter = converter;
            return this;
        }

        /**
         * 设置异常处理对象
         *
         * @param handler
         * @return
         */
        public Builder setExceptionHandler(Cache.ExceptionHandler handler) {
            mExceptionHandler = handler;
            return this;
        }

        /**
         * 设置本地磁盘缓存
         *
         * @param store
         * @return
         */
        public Builder setDiskCacheStore(Cache.CacheStore store) {
            mDiskCacheStore = store;
            return this;
        }

        public CacheConfig build(Context context) {
            mContext = context.getApplicationContext();
            return new CacheConfig(this);
        }
    }
}

File: lib/src/main/java/com/sd/lib/cache/CacheConfig.kt
package com.sd.lib.cache

import android.annotation.SuppressLint
import android.content.Context
import com.sd.lib.cache.store.CacheStore
import com.sd.lib.cache.store.FileCacheStore
import java.io.File
import java.security.MessageDigest

class CacheConfig private constructor(
  builder: Builder,
  context: Context,
) {
  private val context = context.applicationContext
  private val cacheStoreFactory: CacheStoreFactory
  internal val directory: File by lazy { context.filesDir.resolve("sd.lib.cache") }
  internal val objectConverter: ObjectConverter
  internal val exceptionHandler: ExceptionHandler

  init {
    this.cacheStoreFactory = builder.cacheStoreFactory ?: CacheStoreFactory { FileCacheStore() }
    this.objectConverter = builder.objectConverter ?: DefaultObjectConverter()
    this.exceptionHandler = LibExceptionHandler(builder.exceptionHandler)
  }

  /** 创建仓库 */
  internal fun newCacheStore(group: String, id: String): CacheStore? {
    return libRunCatching {
      cacheStoreFactory.create().also { cacheStore ->
        cacheStore.init(
          context = context,
          directory = directory.resolve(md5(group)).resolve(md5(id)),
        )
      }
    }.getOrNull()
  }

  /**
   * 缓存仓库工厂
   */
  fun interface CacheStoreFactory {
    fun create(): CacheStore
  }

  /**
   * 对象转换器，[encode]和[decode]可能存在并发，注意线程安全
   */
  interface ObjectConverter {
    /** 编码 */
    @Throws(Throwable::class)
    fun <T> encode(value: T, clazz: Class<T>): ByteArray

    /** 解码 */
    @Throws(Throwable::class)
    fun <T> decode(bytes: ByteArray, clazz: Class<T>): T
  }

  /**
   * 异常处理类
   */
  fun interface ExceptionHandler {
    fun onException(error: Throwable)
  }

  class Builder {
    internal var cacheStoreFactory: CacheStoreFactory? = null
      private set

    internal var objectConverter: ObjectConverter? = null
      private set

    internal var exceptionHandler: ExceptionHandler? = null
      private set

    /** 缓存仓库工厂 */
    fun setCacheStoreFactory(factory: CacheStoreFactory?) = apply {
      this.cacheStoreFactory = factory
    }

    /** 对象转换 */
    fun setObjectConverter(converter: ObjectConverter) = apply {
      this.objectConverter = converter
    }

    /** 异常处理 */
    fun setExceptionHandler(handler: ExceptionHandler) = apply {
      this.exceptionHandler = handler
    }

    fun build(context: Context): CacheConfig {
      return CacheConfig(this, context)
    }
  }

  companion object {
    @SuppressLint("StaticFieldLeak")
    private var sConfig: CacheConfig? = null

    /** 初始化 */
    @JvmStatic
    fun init(config: CacheConfig) {
      synchronized(this@Companion) {
        if (sConfig == null) {
          sConfig = config
        } else {
          error("CacheConfig has been initialized.")
        }
      }
    }

    internal fun get(): CacheConfig {
      return sConfig ?: synchronized(this@Companion) {
        checkNotNull(sConfig) { "CacheConfig.init() should be called before this." }
      }
    }
  }
}

inline fun CacheConfig.Companion.init(
  context: Context,
  block: CacheConfig.Builder.() -> Unit = {},
) {
  init(
    CacheConfig.Builder()
      .apply(block)
      .build(context)
  )
}

private fun md5(input: String): String {
  val md5Bytes = MessageDigest.getInstance("MD5").digest(input.toByteArray())
  return buildString {
    for (byte in md5Bytes) {
      val hex = (0xff and byte.toInt()).toString(16)
      if (hex.length == 1) append("0")
      append(hex)
    }
  }
}